Solution Homework

Berikut saya lampirkan alternatif pengerjaan homework Python Programming. Semoga bisa menjadi bahan pencerahan untuk teman-teman semua. <b>Tidak ada 1 jawaban pasti </b>, karena seperti yang saya sampaikan di kelas, ada banyak jalan menuju Roma :)

Salam

# Soal 1

Tentukan dari variabel 

list_plat_mobil = ['B 1234 AB', 'B 6721 TY', 'B 1233 AY', 'B 6629 DD', 'B 1111 AM', 'B 6726 D', 'D 11223 KJ', 'AE 44677 GH', 'AE 67269 AA']

Ada berapa jumlah plat genap dan ganjil. 

In [1]:
list_plat_mobil = ['B 1234 AB', 'B 6721 TY', 'B 1233 AY', 'B 6629 DD', 'B 1111 AM', 'B 6726 D', 'D 11223 KJ', 'AE 44677 GH', 'AE 67269 AA']

In [2]:
#buat prototype #1
plat = 'B 1234 AB'
nomor = plat.split() #split membuat setiap item menjadi suatu list
nomor

['B', '1234', 'AB']

In [4]:
#buat prototype #1
plat = 'B 1234 AB'
nomor = plat.split()[1] #split membuat setiap item menjadi suatu list dengan nomor menjadi index 1
nomor = int(nomor)
nomor

1234

In [5]:
n_genap = 0  #inisiasi awal
n_ganjil = 0 #inisiasi awal
for i in list_plat_mobil:
    nomor = int(i.split()[1]) 
    if nomor%2==0:
        n_genap+=1
    else:
        n_ganjil+=1
        
print('Jumlah Plat Genap ada', n_genap)
print('Jumlah Plat Ganjil ada', n_ganjil)

Jumlah Plat Genap ada 2
Jumlah Plat Ganjil ada 7


In [6]:
#buat prototype #2
import re
plat = 'B 1234 AB'
re.findall('\d+', plat) #regex dalam suatu string hanya mengambil numericb

['1234']

In [7]:
#buat prototype #2
plat = 'B 1234 AB'
nomor = re.findall('\d+', plat)[0] #regex dalam suatu string hanya mengambil numeric, karena hanya 1 maka index 0
nomor = int(nomor)
nomor

1234

In [8]:
n_genap = 0  #inisiasi awal
n_ganjil = 0 #inisiasi awal
for i in list_plat_mobil:
    nomor = int(re.findall('\d+', i)[0])
    if nomor%2==0:
        n_genap+=1
    else:
        n_ganjil+=1
        
print('Jumlah Plat Genap ada', n_genap)
print('Jumlah Plat Ganjil ada', n_ganjil)

Jumlah Plat Genap ada 2
Jumlah Plat Ganjil ada 7


# Soal 2

Buatlah deret fibonacci 12 angka yang dimulai dari 2 dan 3 menggunakan iterasi cara for atau while. 

Output yang diharapkan adalah [2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377]

### For Iteration

In [9]:
# identifikasi awal
n1 = 2 #nilai pertama
n2 = 3 #nilai kedua
length = 12
count = 0

list_finbonacci = [n1, n2]
for i in range(length-2): #dilakukan pengurangan 2 karena pada list_finbonacci sudah terisi 2 item n1 dan n2
    nexts = n1 + n2 #untuk membuat variabel nilai selanjutnya dari penjumlahan n1+n2
    # update values
    n1 = n2 #n1 menjadi n2
    n2 = nexts #n2 menjadi bilangan selanjutnya atau nexts
    list_finbonacci.append(nexts) #melakukan penambahan item pada list setiap iterasi

print(list_finbonacci)

[2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377]


### While Iteration

In [10]:
# identifikasi awal
n1 = 2 #nilai pertama
n2 = 3 #nilai kedua
length = 12
count = 0

list_finbonacci = [n1, n2]
while count < length-2: #dilakukan pengurangan 2 karena pada list_finbonacci sudah terisi 2 item n1 dan n2
    nexts = n1 + n2 #untuk membuat variabel nilai selanjutnya dari penjumlahan n1+n2
    # update values
    n1 = n2
    n2 = nexts
    list_finbonacci.append(nexts) #melakukan penambahan item pada list setiap iterasi
    count += 1 

print(list_finbonacci)

[2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377]


# Soal 3

Team Campaign PT ABC akan membuat promo bundling category product (berisi 3 category setiap bundling) untuk promo 17 Agustus.  PT ABC memiliki 10 category product.  Ketentuan Bundling tidak boleh hanya memiliki 1 jenis category dalam 1 bundling dan urutan tidak menjadi faktor pembeda (c1 - c2 - c3 = c2 - c1 - c3). Tentukan ada berapa jumlah kemungkinan bundling yang bisa terbentuk?

Berikut list code category <b>list_id_category</b> = [c1, c2, c3, c4, c5, c6, c7 , c8, c9, c10]

In [11]:
list_id_category = ['c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7' , 'c8', 'c9', 'c10']

In [12]:
sorted(['a1', 'a3', 'a2'])

['a1', 'a2', 'a3']

In [13]:
list_bundling_category = []

for i in list_id_category:
    a1 = i #category 1
    for y in list_id_category:
        a2 = y #category 2
        for x in list_id_category:
            a3 = x #category 3
            list_temp = sorted([a1, a2, a3]) #perlu di sort terlebih dulu dibuat list temporary yang selalu berubah
            if list_temp not in list_bundling_category and a1!=a2!=a3:
                list_bundling_category.append(list_temp)
                
# print(list_bundling_category)
print('Terdapat', len(list_bundling_category), 'kemungkinan bundling category untuk keperluan campaign')

Terdapat 210 kemungkinan bundling category untuk keperluan campaign


# Soal 4

Buatlah sebuah function string_processor dengan inputan 2 parameter, yaitu suatu string 1 kata dan nama proses yang akan dilakukan. Function ini memiliki 4 proses
- first_char : return huruf pertama dari string yang diinput
- last_char : return huruf terakhir dari string yang diinput
- num_chars : return jumlah huruf string input
- num_vowels : return jumlah huruf vocal pada string input


In [14]:
def string_processor(metode, string):
    string = string.lower()
    if metode == 'first_char':
        result = string[0]
    elif metode == 'last_char':
        result = string[-1]
    elif metode == 'num_chars':
        result = len(string)
    elif metode == 'num_vowels':
        num_vowels = 0
        for i in string:
          if i in ['a', 'e', 'i', 'u', 'o']:
            num_vowels += 1
        result = num_vowels
    else:
        result = 'Inputan Salah'
    return result

In [15]:
#TESTING

t1 = 'pasar'
t2 = 'rumah'
t3 = 'mawar'
t4 = 'ibadah'

print('Hasil Testing first char dari',t1, 'adalah', string_processor('first_char', t1))
print('Hasil Testing last char dari',t2, 'adalah', string_processor('last_char', t2))
print('Hasil Testing total chars dari',t3, 'adalah', string_processor('num_chars', t3))
print('Hasil Testing total chars vowel dari',t4, 'adalah', string_processor('num_vowels', t4))

Hasil Testing first char dari pasar adalah p
Hasil Testing last char dari rumah adalah h
Hasil Testing total chars dari mawar adalah 5
Hasil Testing total chars vowel dari ibadah adalah 3


# Soal 5

Menggunakan dataset rakamin_store_city_monthly! Buatlah 1 kolom baru pada dengan teknik apply lambda tanpa function def (cukup one liner) dengan nama kolom is_gmail dengan value True jika menggunakan gmail dan False jika selain gmail. 

Tentukan ada berapa user yang menggunakan email gmail pada tahun 2018!

In [16]:
import pandas as pd

df = pd.read_csv('rakamin_store_city_monthly.csv', sep=';')
df.head()

month  user_id                 username  \
0  2016-01       16            Kevin Ningsih   
1  2016-01       25            Sakti Hasaniy   
2  2016-01       39              Izhar Halim   
3  2016-02        8  Dikposa Yosua Krisnanto   
4  2016-02       34          Ismail Wardhani   

                               email             kota  total_visit  \
0         kevinningsih@roketmail.com  Jakarta Selatan           95   
1           saktihasaniy@outlook.com    Jakarta Pusat           97   
2               izharhalim@gmail.com    Jakarta Timur           84   
3  dikposayosuakrisnanto@hotmail.com    Jakarta Pusat          113   
4         ismailwardhani@hotmail.com  Jakarta Selatan           25   

   total_order  total_amount  
0           15         45724  
1            1         26665  
2           33        899001  
3           11         22012  
4            2         55620

In [17]:
#get year
df['year'] = df['month'].apply(lambda x: x[:4])
df['year'] = df['year'].astype(int)
df.head()

month  user_id                 username  \
0  2016-01       16            Kevin Ningsih   
1  2016-01       25            Sakti Hasaniy   
2  2016-01       39              Izhar Halim   
3  2016-02        8  Dikposa Yosua Krisnanto   
4  2016-02       34          Ismail Wardhani   

                               email             kota  total_visit  \
0         kevinningsih@roketmail.com  Jakarta Selatan           95   
1           saktihasaniy@outlook.com    Jakarta Pusat           97   
2               izharhalim@gmail.com    Jakarta Timur           84   
3  dikposayosuakrisnanto@hotmail.com    Jakarta Pusat          113   
4         ismailwardhani@hotmail.com  Jakarta Selatan           25   

   total_order  total_amount  year  
0           15         45724  2016  
1            1         26665  2016  
2           33        899001  2016  
3           11         22012  2016  
4            2         55620  2016

In [18]:
#prototype logic gmail or not
email = 'izharhalim@gmail.com'
'gmail' in email #check true / false menggunakan in

True

In [19]:
df['is_gmail'] = df['email'].apply(lambda x: 'gmail' in x)
df.head()

month  user_id                 username  \
0  2016-01       16            Kevin Ningsih   
1  2016-01       25            Sakti Hasaniy   
2  2016-01       39              Izhar Halim   
3  2016-02        8  Dikposa Yosua Krisnanto   
4  2016-02       34          Ismail Wardhani   

                               email             kota  total_visit  \
0         kevinningsih@roketmail.com  Jakarta Selatan           95   
1           saktihasaniy@outlook.com    Jakarta Pusat           97   
2               izharhalim@gmail.com    Jakarta Timur           84   
3  dikposayosuakrisnanto@hotmail.com    Jakarta Pusat          113   
4         ismailwardhani@hotmail.com  Jakarta Selatan           25   

   total_order  total_amount  year  is_gmail  
0           15         45724  2016     False  
1            1         26665  2016     False  
2           33        899001  2016      True  
3           11         22012  2016     False  
4            2         55620  2016     False

In [20]:
# Filter tahun 2018

df_year = df[(df.year == 2018)].groupby(['year', 'is_gmail'])['user_id'].nunique().reset_index()
df_year.columns = ['year', 'is_gmail', 'total_user']
df_year

year  is_gmail  total_user
0  2018     False          14
1  2018      True           3

In [21]:
n_true = df_year[(df_year.is_gmail == True)].reset_index()['total_user'][0]

In [22]:
print('Total user yang menggunakan email gmail pada tahun 2018 sebanyak', n_true)

Total user yang menggunakan email gmail pada tahun 2018 sebanyak 3


# Soal 6

Menggunakan dataset superstore! Buatlah function yang bisa menghitung Jumlah Sales dan Profit pada suatu tahun dan region tertentu dengan default tahun 2017. Hint : Buat kolom tahun

<b>Tentukan total sales dan profit:</b><br>
- South Region pada tahun 2016
- West Region pada tahun 2017

In [23]:
import pandas as pd
import numpy as np

df = pd.read_excel('superstore.xls')
df.head(3)

Row ID        Order ID Order Date  Ship Date     Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16  Second Class    DV-13045   

     Customer Name    Segment        Country         City  ... Postal Code  \
0      Claire Gute   Consumer  United States    Henderson  ...       42420   
1      Claire Gute   Consumer  United States    Henderson  ...       42420   
2  Darrin Van Huff  Corporate  United States  Los Angeles  ...       90036   

   Region       Product ID         Category Sub-Category  \
0   South  FUR-BO-10001798        Furniture    Bookcases   
1   South  FUR-CH-10000454        Furniture       Chairs   
2    West  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.96         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.62         2   

   Discount    Profit  
0       0.0   41.9136  
1       0.0  219.5820  
2       0.0    6.8714  

[3 rows x 21 columns]

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9994 non-null   int64         
 1   Order ID       9994 non-null   object        
 2   Order Date     9994 non-null   datetime64[ns]
 3   Ship Date      9994 non-null   datetime64[ns]
 4   Ship Mode      9994 non-null   object        
 5   Customer ID    9994 non-null   object        
 6   Customer Name  9994 non-null   object        
 7   Segment        9994 non-null   object        
 8   Country        9994 non-null   object        
 9   City           9994 non-null   object        
 10  State          9994 non-null   object        
 11  Postal Code    9994 non-null   int64         
 12  Region         9994 non-null   object        
 13  Product ID     9994 non-null   object        
 14  Category       9994 non-null   object        
 15  Sub-Category   9994 n

In [25]:
#get year
df['year'] = df['Order Date'].dt.year #karena type data Order Date Sudah date time (datetime64[ns]) bisa menggunakan fungsi .dt
df[['year','Order Date','Order ID', 'Region', 'Sales', 'Profit']].head(3)

year Order Date        Order ID Region   Sales    Profit
0  2016 2016-11-08  CA-2016-152156  South  261.96   41.9136
1  2016 2016-11-08  CA-2016-152156  South  731.94  219.5820
2  2016 2016-06-12  CA-2016-138688   West   14.62    6.8714

In [26]:
def total_sales_profit(region, year=2017): #default year 2017
    #filter dataframe
    df_filter = df[(df.year == year) & (df.Region == region)]

    #get value sum sales and profit
    total_sales = df_filter.Sales.sum()
    total_profit = df_filter.Profit.sum()

    #return statement 2 values
    return total_sales, total_profit #tuple

In [27]:
#pertanyaan 1
print('Total Sales pada South Region tahun 2016 adalah', round(total_sales_profit('South', 2016)[0], 2))
print('Total Profit pada South Region tahun 2016 adalah', round(total_sales_profit('South', 2016)[1], 2))

Total Sales pada South Region tahun 2016 adalah 93610.22
Total Profit pada South Region tahun 2016 adalah 17702.81


In [28]:
#pertanyaan 2 v1
print('Total Sales pada West Region tahun 2017 adalah', round(total_sales_profit('West', 2017)[0], 2))
print('Total Profit pada West Region tahun 2017 adalah', round(total_sales_profit('West', 2017)[1], 2))

Total Sales pada West Region tahun 2017 adalah 250128.37
Total Profit pada West Region tahun 2017 adalah 43808.96


In [29]:
#pertanyaan 2 v2
print('Total Sales pada West Region tahun 2017 adalah', round(total_sales_profit('West')[0], 2))
print('Total Profit pada West Region tahun 2017 adalah', round(total_sales_profit('West')[1], 2))

Total Sales pada West Region tahun 2017 adalah 250128.37
Total Profit pada West Region tahun 2017 adalah 43808.96


# Soal 7

Menggunakan dataset superstore! Buatlah 3 kelompok 'segment'  (high, medium, dan low) berdasarkan performance pembelian yang digunakan sebagai acuan tim bisnis untuk mengelompokan customer superstore. <br>

<b>Instruksi:</b>
1. Bantulah tim bisnis untuk menentukan kriteria pengelompokan untuk 3 segmentasi customer ini, dan gunakan iterrows atau apply lambda sebagai teknik pembuatan kolom baru!
2. Berikan alasan kalian dalam penentuan kriteria pengelompokan customer yang dibuat! 
3. Tentukan jumlah customer pada masing-masing segment di tahun 2017! 

Hint : Pahami setiap row pada dataset, pastikan jumlah customer unique


In [30]:
import pandas as pd

df = pd.read_excel('superstore.xls')
df.head(3)

Row ID        Order ID Order Date  Ship Date     Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16  Second Class    DV-13045   

     Customer Name    Segment        Country         City  ... Postal Code  \
0      Claire Gute   Consumer  United States    Henderson  ...       42420   
1      Claire Gute   Consumer  United States    Henderson  ...       42420   
2  Darrin Van Huff  Corporate  United States  Los Angeles  ...       90036   

   Region       Product ID         Category Sub-Category  \
0   South  FUR-BO-10001798        Furniture    Bookcases   
1   South  FUR-CH-10000454        Furniture       Chairs   
2    West  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.96         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.62         2   

   Discount    Profit  
0       0.0   41.9136  
1       0.0  219.5820  
2       0.0    6.8714  

[3 rows x 21 columns]

In [31]:
#get year
#get year
df['year'] = df['Order Date'].dt.year #karena type data Order Date Sudah date time (datetime64[ns]) bisa menggunakan fungsi .dt

Notes:
- 1 row berisi 1 order-code dari detail pembelian seorang customer
- Max order date 30 Desember 2017
- Historical data 4 tahun, dari 2014 - 2017
- Historical data 4 region (south, west, east and central)
- Metric yang kami gunakan untuk melakukan segmentasi menggunakan performance sales

In [32]:
df['Order Date'].max() #max date

Timestamp('2017-12-30 00:00:00')

In [33]:
#persebaran order per Year
df['year'].value_counts()

2017    3312
2016    2587
2015    2102
2014    1993
Name: year, dtype: int64

In [34]:
#persebaran order per Region
df['Region'].value_counts()

West       3203
East       2848
Central    2323
South      1620
Name: Region, dtype: int64

In [35]:
#Karena kita ingin melakukan segmentasi customer dan dataset masih dalam level order ID maka kita perlu
#mengagregatnya terlebih dulu untuk mendapatkan performance sales pada level customer_id

df_group_year = df.groupby(['year','Customer ID', 'Customer Name'])['Sales'].sum().reset_index()

df_group_year.head()

year Customer ID  Customer Name     Sales
0  2014    AA-10315     Alex Avila   756.048
1  2014    AA-10375   Allen Armold    50.792
2  2014    AA-10480   Andrew Allen    27.460
3  2014    AA-10645  Anna Andreadi  1434.330
4  2014    AB-10015  Aaron Bergman   322.216

In [36]:
#check distribution statistic descriptive sales setiap user per tahunnya

df_group_year_desc = df_group_year.groupby(['year'])['Sales'].describe(percentiles = [.25, .5, .75, .9, .95]).reset_index()
df_group_year_desc

year  count         mean          std    min      25%      50%        75%  \
0  2014  595.0   813.861341  1558.339428  1.869   86.448  321.552   951.7495   
1  2015  573.0   821.173663  1115.847785  1.720  120.564  389.434  1065.9200   
2  2016  638.0   954.867708  1524.135920  1.964  107.271  495.729  1210.8170   
3  2017  693.0  1058.030671  1463.910713  1.188  204.405  604.910  1334.6416   

         90%         95%        max  
0  2016.9944  2764.38828  23661.228  
1  2210.6500  2900.66280   9022.324  
2  2090.1135  3216.82335  18344.052  
3  2384.6104  3755.71360  14203.278

Dari hasil statistic descriptive di atas kita mengatahui bahwa nilai median (p50) setiap tahun nya bernilai <b>fluktuatif</b>. Sehingga dari hasil ini kami memutuskan untuk membuat definisi segmentasi performance pembelian akan berbeda setiap tahunnya menggunakan pendekatan statistic descriptive pada performance sales setiap tahunnya. <br>

Selanjutnya, untuk membagi threshold setiap segment, kami menggunakan acuan percentile.
1. low --> total sales per tahun < p50
2. mid --> total sales per tahun >= p50 and <= 75
3. high --> total sales per tahun > p75

In [37]:
df_group_year = df_group_year.merge(df_group_year_desc[['year', 'mean', '25%', '50%', '75%']], on='year', how='inner')
df_group_year.head()

year Customer ID  Customer Name     Sales        mean     25%      50%  \
0  2014    AA-10315     Alex Avila   756.048  813.861341  86.448  321.552   
1  2014    AA-10375   Allen Armold    50.792  813.861341  86.448  321.552   
2  2014    AA-10480   Andrew Allen    27.460  813.861341  86.448  321.552   
3  2014    AA-10645  Anna Andreadi  1434.330  813.861341  86.448  321.552   
4  2014    AB-10015  Aaron Bergman   322.216  813.861341  86.448  321.552   

        75%  
0  951.7495  
1  951.7495  
2  951.7495  
3  951.7495  
4  951.7495

In [38]:
#function segmentation user besed on total sales yearly
def segment(x):
    if x['Sales'] < x['50%']:
        segment = 'Low'
    elif x['Sales'] <= x['75%'] and x['Sales'] >= x['50%']:
        segment = 'Mid'
    else:
        segment = 'High'
    return segment

In [39]:
df_group_year['segment'] = df_group_year.apply(lambda x: segment(x), axis=1)
df_group_year.head()

year Customer ID  Customer Name     Sales        mean     25%      50%  \
0  2014    AA-10315     Alex Avila   756.048  813.861341  86.448  321.552   
1  2014    AA-10375   Allen Armold    50.792  813.861341  86.448  321.552   
2  2014    AA-10480   Andrew Allen    27.460  813.861341  86.448  321.552   
3  2014    AA-10645  Anna Andreadi  1434.330  813.861341  86.448  321.552   
4  2014    AB-10015  Aaron Bergman   322.216  813.861341  86.448  321.552   

        75% segment  
0  951.7495     Mid  
1  951.7495     Low  
2  951.7495     Low  
3  951.7495    High  
4  951.7495     Mid

In [40]:
#filter tahun 2017 dan count unique based on customer id
df_group_year[(df_group_year.year == 2017)].groupby(['segment'])['Customer ID'].nunique()

segment
High    173
Low     346
Mid     174
Name: Customer ID, dtype: int64

*Pada* tahun 2017, ~25% atau sekitar 173 dari total customer kita masuk dalam ketegori High

**Tetap Semangat Belajat Python! Practice Make Perfect!**